src/coreclr/jit/codegenwasm.cpp

```diff
@@ -59,6 +121,253 @@ void CodeGen::genFuncletEpilog()
+//
+// Arguments:
+//   block - block to finish up
+//
+// Returns:
+//   Updated block to process (if not block->Next()) or nullptr.
+//
+BasicBlock* CodeGen::genEmitEndBlock(BasicBlock* block)
+{
+    // Generate Wasm control flow instructions for block ends
+    //
+    switch (block->GetKind())
+    {
+        case BBJ_RETURN:
+            genExitCode(block);
+            break;
+
+        case BBJ_THROW:
+            instGen(INS_unreachable);
+            break;
+
+        case BBJ_CALLFINALLY:
+            // Wasm-TODO: EH model
+            genCallFinally(block);
+            if (!block->isBBCallFinallyPair())
+            {
+                instGen(INS_unreachable);
+            }
+            break;
+
+        case BBJ_EHCATCHRET:
+            // Wasm-TODO: EH model
+            assert(compiler->UsesFunclets());
+            genEHCatchRet(block);
+            FALLTHROUGH;
+
+        case BBJ_EHFINALLYRET:
+        case BBJ_EHFAULTRET:
+        case BBJ_EHFILTERRET:
+            // Wasm-TODO: EH model
+            if (compiler->UsesFunclets())
+            {
+                genReserveFuncletEpilog(block);
+            }
+            instGen(INS_return);
+            break;
+
+        case BBJ_SWITCH:
+        {
+            BBswtDesc* const desc      = block->GetSwitchTargets();
+            unsigned const   caseCount = desc->GetCaseCount();
+
+            assert(!desc->HasDefaultCase());
+
+            if (caseCount == 0)
+            {
+                break;
+            }
+
+            inst_SWITCH(caseCount);
+
+            for (unsigned caseNum = 0; caseNum < caseCount; caseNum++)
+            {
+                BasicBlock* const caseTarget = desc->GetCase(caseNum)->getDestinationBlock();
+                unsigned          depth      = findTargetDepth(caseTarget);
+                inst_LABEL(depth);
+            }
+
+            JITDUMP("\n");
+            break;
+        }
+
+        case BBJ_CALLFINALLYRET:
+        case BBJ_ALWAYS:
+        {
+#ifdef DEBUG
+            GenTree* call = block->lastNode();
+            if ((call != nullptr) && call->OperIs(GT_CALL))
+            {
+                // At this point, BBJ_ALWAYS should never end with a call that doesn't return.
+                assert(!call->AsCall()->IsNoReturn());
+            }
+#endif // DEBUG
+
+            BasicBlock* const target = block->GetTarget();
+
+            // If this block jumps to the next one, we might be able to skip emitting the jump
+            //
+            if (block->CanRemoveJumpToNext(compiler))
+            {
+                assert(getBlockIndex(target) == (getBlockIndex(block) + 1));
+                break;
+            }
+
+            inst_JMP(EJ_jmp, target);
+            break;
+        }
+
+        case BBJ_COND:
+        {
+            BasicBlock* const trueTarget  = block->GetTrueTarget();
+            BasicBlock* const falseTarget = block->GetFalseTarget();
+
+            // We don't expect degenerate BBJ_COND
```

Comment:
What prevents this?

Reply:
Going to leave this as is, though it seems like something we can trivially handle.